src/types/check.rs

```diff
@@ -60,7 +60,8 @@ impl PrimType {
                 Ok(())
             }
             (PrimType::Function(t), PrimValue::Function { hugr: v })
-                if Some(t.as_ref()) == v.get_function_type() =>
+                if t.params.is_empty() // Hugr's still monomorphic
+                && Some(&t.body) == v.get_function_type() =>
```

Comment:
Is this `is_empty` check necessary? I would presume `v.get_function_type()` will have empty params, so the equality check should take care of it

Reply:
1. At the moment `v.get_function_type()` returns a `&FunctionType`, not a `PolyFuncType`. I can't map that into a no-binders PolyFuncType without `clone`ing it, which seems a bit wasteful (but you if say "premature optimization" I'm willing to change that I guess).
2. I could add an `PolyFuncType::eq_monomorphic(&FunctionType)`, which is a little nicer in some ways, but...
3. In fact `v.get_function_type()` refers an *Option* of `&FunctionType` and by the time I've dealt with that, basically *anything else* is 3x longer for unclear benefit

Other suggestions welcome!

Reply:
Ah, I missed that there is Option::is_some_and. So we could add `PolyFuncType::eq_monomorphic(&self, &FunctionType)` and then make this here `v.get_function_type().is_some_and(|f|t.eq_monomorphic(f))`, that might be a bit better?

Reply:
Ok, and rather than `eq_monomorphic`, the trait "PartialEq" is parametrised, so that does look like the right way to do this! Done :)

Reply:
Done!

Reply:
Very nice!